internal/pkg/controllers/binding/binding.go

```diff
@@ -0,0 +1,260 @@
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/security-profiles-operator/api/profilebinding/v1alpha1"
+)
+
+const (
+	reconcileTimeout = 1 * time.Minute
+
+	errGetBinding = "cannot get binding"
+
+	reasonCannotUpdateProfileBinding       event.Reason = "CannotUpdateProfileBinding"
+	reasonCannotUpdateProfileBindingStatus event.Reason = "CannotUpdateProfileBindingStatus"
+	reasonCannotRetrieveProfileBinding     event.Reason = "CannotRetrieveProfileBinding"
+
+	profileRefKey = ".metadata.profileRef"
+	boundPodsKey  = ".metadata.boundPods"
+
+	finalizer = "active-workload-lock"
+)
+
+// Setup adds a controller that reconciles profile bindings.
+func Setup(ctx context.Context, mgr ctrl.Manager, l logr.Logger) error {
+	// Index ProfileBindings by the Profile they reference
+	if err := mgr.GetFieldIndexer().IndexField(
+		ctx, &v1alpha1.ProfileBinding{}, profileRefKey, func(rawObj runtime.Object) []string {
+			pb, ok := rawObj.(*v1alpha1.ProfileBinding)
+			if !ok {
+				return []string{}
+			}
+			return []string{pb.GetNamespace() + "/" + pb.Spec.Profile}
+		}); err != nil {
+		return errors.Wrap(err, "creating profile index on profilebinding")
+	}
+	// Index ProfileBindings by their active workloads
+	if err := mgr.GetFieldIndexer().IndexField(
+		ctx, &v1alpha1.ProfileBinding{}, boundPodsKey, func(rawObj runtime.Object) []string {
+			pb, ok := rawObj.(*v1alpha1.ProfileBinding)
+			if !ok {
+				return []string{}
+			}
+			return pb.Status.ActiveWorkloads
+		}); err != nil {
+		return errors.Wrap(err, "creating pod index on profilebinding")
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("pods").
+		For(&corev1.Pod{}).
+		WithEventFilter(resource.NewPredicates(hasSeccompProfile)).
+		Complete(&Reconciler{
+			client: mgr.GetClient(),
+			log:    l,
+			record: event.NewAPIRecorder(mgr.GetEventRecorderFor("profilebinding")),
+		})
+}
+
+type Reconciler struct {
+	client client.Client
+	log    logr.Logger
+	record event.Recorder
+}
+
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	logger := r.log.WithValues("pod", req.Name, "namespace", req.Namespace)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	podID := req.Namespace + "/" + req.Name
+
+	pod := &corev1.Pod{}
+	var err error
+	if err = r.client.Get(ctx, req.NamespacedName, pod); ignoreNotFound(err) != nil {
+		return reconcile.Result{}, errors.Wrap(ignoreNotFound(err), errGetBinding)
+	}
+	if kerrors.IsNotFound(err) { // this is a pod deletion, so update all seccomp profiles that were using it
+		return reconcile.Result{}, r.removeBoundPod(ctx, logger, podID)
```

Comment:
I'm feeling a little shaky about this bit. the client we're using is a cached client, and so our references to objects might be a little old. Perhaps this is something better handled in a webhook?